File: sandbox/backend/java/simple-eccomerce/src/main/java/com/simple/ecommerce/application/exception/ProductNotFoundException.java
package com.simple.ecommerce.application.exception;

/**
 * Exception thrown when a product cannot be found.
 */
public class ProductNotFoundException extends RuntimeException {
    
    public ProductNotFoundException(String message) {
        super(message);
    }
    
    public ProductNotFoundException(String message, Throwable cause) {
        super(message, cause);
    }
    
    public ProductNotFoundException(Long productId) {
        super("Product with ID " + productId + " not found");
    }
    
    public ProductNotFoundException(String productName, String additionalInfo) {
        super("Product with name '" + productName + "' not found. " + additionalInfo);
    }
} 

File: Projects/Testtrack/p911 - Simple eCommerce backend/application/exception/product_not_found_exception.py
from typing import Optional, Union


class ProductNotFoundException(Exception):
    def __init__(self, message_or_id: Union[str, int],
                 additional_info_or_cause: Optional[Union[str,
                                                          Exception]] = None,
                 cause: Optional[Exception] = None):
        if isinstance(message_or_id, int):
            message = f"Product with ID {message_or_id} not found"
            super().__init__(message)
            self.__cause__ = additional_info_or_cause if isinstance(
                additional_info_or_cause, Exception) else cause
        elif isinstance(additional_info_or_cause, str):
            message = f"Product with name '{message_or_id}' not found. {additional_info_or_cause}"
            super().__init__(message)
            self.__cause__ = cause
        else:
            super().__init__(message_or_id)
            self.__cause__ = additional_info_or_cause if isinstance(
                additional_info_or_cause, Exception) else cause
